extensions/eclipse-che-theia-plugin-ext/src/plugin/che-sidecar-file-system.ts

```diff
@@ -0,0 +1,223 @@
+/**********************************************************************
+ * Copyright (c) 2018-2020 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+import { CheSideCarFileSystem, FileTypeMain, PLUGIN_RPC_CONTEXT } from '../common/che-protocol';
+import { Stats, lstat, readFile, readdir, stat } from 'fs';
+
+import { RPCProtocol } from '@theia/plugin-ext/lib/common/rpc-protocol';
+import { URI } from 'vscode-uri';
+import { promisify } from 'util';
+
+export namespace SideCarFileSystemProvider {
+  export interface StatAndLink {
+    // The stats of the file. If the file is a symbolic
+    // link, the stats will be of that target file and
+    // not the link itself.
+    // If the file is a symbolic link pointing to a non
+    // existing file, the stat will be of the link and
+    // the `dangling` flag will indicate this.
+    stat: Stats;
+
+    // Will be provided if the resource is a symbolic link
+    // on disk. Use the `dangling` flag to find out if it
+    // points to a resource that does not exist on disk.
+    symbolicLink?: { dangling: boolean };
+  }
+}
+
+export enum FileSystemProviderErrorCode {
+  FileExists = 'EntryExists',
+  FileNotFound = 'EntryNotFound',
+  FileNotADirectory = 'EntryNotADirectory',
+  FileIsADirectory = 'EntryIsADirectory',
+  FileExceedsMemoryLimit = 'EntryExceedsMemoryLimit',
+  FileTooLarge = 'EntryTooLarge',
+  NoPermissions = 'NoPermissions',
+  Unavailable = 'Unavailable',
+  Unknown = 'Unknown',
+}
+
+export class FileSystemProviderError extends Error {
+  constructor(message: string, public readonly code: FileSystemProviderErrorCode) {
+    super(message);
+    Object.setPrototypeOf(this, FileSystemProviderError.prototype);
+  }
+}
+
+export function markAsFileSystemProviderError(error: Error, code: FileSystemProviderErrorCode): Error {
+  error.name = code ? `${code} (FileSystemError)` : 'FileSystemError';
+
+  return error;
+}
+
+export function createFileSystemProviderError(
+  error: Error | string,
+  code: FileSystemProviderErrorCode
+): FileSystemProviderError {
+  const providerError = new FileSystemProviderError(error.toString(), code);
+  markAsFileSystemProviderError(providerError, code);
+
+  return providerError;
+}
+
+export class CheSideCarFileSystemImpl implements CheSideCarFileSystem {
+  constructor(rpc: RPCProtocol) {
+    const delegate = rpc.getProxy(PLUGIN_RPC_CONTEXT.CHE_SIDECAR_FILE_SYSTEM_MAIN);
+    const machineName = process.env.CHE_MACHINE_NAME;
+    if (machineName) {
+      delegate.$registerFileSystemProvider(`file-sidecar-${machineName}`);
```

Comment:
I was worried that this line would suffer from https://github.com/eclipse-theia/theia/issues/8429, but since it's an extension API, it is initialized later than the vscode/theia api and does not have timing problems.